types/jest/index.d.ts

```diff
@@ -389,18 +389,18 @@ declare namespace jest {
         : Value extends Func
         ? SpyInstance<ReturnType<Value>, ArgsType<Value>>
         : never;
-    function spyOn<T extends {}, M extends FunctionPropertyNames<Required<T>>>(
-        object: T,
-        method: M,
-    ): FunctionProperties<Required<T>>[M] extends Func
-        ? SpyInstance<ReturnType<FunctionProperties<Required<T>>[M]>, ArgsType<FunctionProperties<Required<T>>[M]>>
-        : never;
     function spyOn<T extends {}, M extends ConstructorPropertyNames<Required<T>>>(
         object: T,
         method: M,
     ): Required<T>[M] extends new (...args: any[]) => any
         ? SpyInstance<InstanceType<Required<T>[M]>, ConstructorArgsType<Required<T>[M]>>
         : never;
+    function spyOn<T extends {}, M extends FunctionPropertyNames<Required<T>>>(
```

Comment:
Swapping the order here breaks this use case:

```typescript
jest.spyOn(Storage.prototype, 'getItem').mockImplementation((key) => null);
```

undoing this change breaks

```typescript
let typedSpy1: jest.SpiedClass<typeof globalThis.Date>;
typedSpy1 = jest.spyOn(globalThis, 'Date');
```

How can we make both work?

Reply:
See #63182. Seems like it will make both cases work as expected.

Reply:
Opened #63196 to address this.

Reply:
Missed that one, thanks!

Reply:
Just to add. As you noticed swapping the order made `spyOn` work correctly with `Date`. The type of `Date` is rather odd indeed, but the same could be said about `Storage`. Not so easy to cover both.